### PR TITLE
PWX-37369 Automation checking for pod name in pxctl v i

### PR DESCRIPTION
**Pull request by sayalasomayajula-px**
**What this PR does / why we need it**:
Checks the pod name is there in the volume labels of the volume which is carved out from the storage class which has fa_pod_name as a parameter

**Which issue(s) this PR fixes** (optional)
Closes #
PWX-37369
**Special notes for your reviewer**:

PASS LOGS: (https://jenkins.pwx.dev.purestorage.com/job/Users/job/krishna/job/tp-pxe-byoc/3327/display/redirect)
https://jenkins.pwx.dev.purestorage.com/job/Users/job/krishna/job/tp-pxe-byoc/3278/
Tested with 1 repl, as testbed is too small for the test

### CODE REVIEW

File: drivers/scheduler/k8s/specs/fio-fada-realm/px-storage-class.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: fio-realm-pod-sc
+provisioner: pxd.portworx.com
+parameters:
+  backend: "pure_block"
+  max_iops: "1000"
+  max_bandwidth: "1G"
+  fs: "ext4"
+  pure_fa_pod_name: "firstpod"
```
Comment (vumesh-px) [inline drivers/scheduler/k8s/specs/fio-fada-realm/px-storage-class.yaml:11]: this cannot be hard-coded, needs to be placeholder and we will need to create a new pod and pass it

File: drivers/scheduler/k8s/specs/fio-fada-realm/px-storage-class.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: fio-log-realm-pod-sc
+provisioner: pxd.portworx.com
+parameters:
+  backend: "pure_block"
+  max_iops: "1000"
+  max_bandwidth: "1G"
+  fs: "ext4"
+  pure_fa_pod_name: "firstpod"
```
Comment (vumesh-px) [inline drivers/scheduler/k8s/specs/fio-fada-realm/px-storage-class.yaml:27]: same as above

File: tests/basic/pure_test.go [hunk 3]
```diff
@@ -5002,3 +5002,87 @@ var _ = Describe("{RebootingNodesWhileFADAvolumeCreationInProgressUsingZones}",
 		AfterEachTest(contexts)
 	})
 })
+
+var _ = Describe("{ValidatePodNameinVolume}", func() {
+	/*
+		https://purestorage.atlassian.net/browse/PWX-37369
+		As part of Mulitenancy feature, we are deploying a app which has "pure_fa_pod_name" in the storage class
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5009]:
```suggestion
		As part of Mulitenancy feature, we are deploying an app which has "pure_fa_pod_name" in the storage class
```

File: tests/basic/pure_test.go [hunk 4]
```diff
@@ -5002,3 +5002,87 @@ var _ = Describe("{RebootingNodesWhileFADAvolumeCreationInProgressUsingZones}",
...
+		1. Deploy an app which uses FADA volume( uses csi provisioner and a pure_block backend) and validate the pod name in the volume
+		2. Check if the pod name in the volume is same as the pod name in the storage class
+		3.If the pod name in the volume is not same as the pod name in the storage class, fail the test and even fail the t
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5012]: incomplete sentence

**Review by vumesh-px — CHANGES_REQUESTED** [review]
Test needs to create a pod in the FA mentioned (px-pure-secret) and then pass that to the storageclass

File: tests/basic/pure_test.go [hunk 5]
```diff
@@ -5002,3 +5004,136 @@ var _ = Describe("{RebootingNodesWhileFADAvolumeCreationInProgressUsingZones}",
...
+		if !isRealmExists {
+			log.FailOnError(fmt.Errorf("Realm [%v] is not created in FA", RealmName), "is realm created in FA?")
+		}
+		podNameinSC := "Torpedo-Test" + Inst().InstanceID
+		PodNameinFA := RealmName + "::" + podNameinSC
+
+		stepLog := "Create A pod inside Realm"
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			_, err = pureutils.CreatePodinFA(faClient, PodNameinFA)
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5059]: do you first want to check if the pod already exists before creating?

File: tests/basic/pure_test.go [hunk 6]
```diff
@@ -5002,3 +5004,136 @@ var _ = Describe("{RebootingNodesWhileFADAvolumeCreationInProgressUsingZones}",
...
+				AppKeys:            Inst().AppList,
+				StorageProvisioner: fmt.Sprintf("%v", portworx.PortworxCsi),
+				Namespace:          testName,
+			})
+			log.FailOnError(err, "Failed to schedule application of %v namespace", testName)
+			contexts = append(contexts, context...)
+			ValidateApplications(contexts)
+
+		})
+		stepLog = "Fetch the pod name from the storage class"
```
Comment (vumesh-px) [inline tests/basic/pure_test.go:5099]: do you mean "Fetch the pod name from the volume"?

**Review by vumesh-px — CHANGES_REQUESTED** [review]

File: tests/basic/pure_test.go (hunk 5 above)
Comment (sayalasomayajula-px) [reply tests/basic/pure_test.go]: Added it

File: tests/basic/pure_test.go (hunk 6 above)
Comment (sayalasomayajula-px) [reply tests/basic/pure_test.go]: Added the logging

**Review by vumesh-px — APPROVED** [review]
LGTM

File: tests/basic/pure_test.go [hunk 7]
```diff
@@ -5002,3 +5004,140 @@ var _ = Describe("{RebootingNodesWhileFADAvolumeCreationInProgressUsingZones}",
...
+				volumes, err := Inst().S.GetVolumes(ctx)
+				log.FailOnError(err, "Failed to get volumes")
+				for _, volume := range volumes {
+					inspectedVolume, err := Inst().V.InspectVolume(volume.ID)
+					log.FailOnError(err, "Failed to inspect volume")
+					PodName := inspectedVolume.Locator.VolumeLabels["pure_fa_pod_name"]
+					if PodName == "" {
+						log.FailOnError(fmt.Errorf("Pod Name is not present in the volume labels"), "is pod name present in volume labels?")
+					} else if PodName != podNameinSC {
+						log.FailOnError(fmt.Errorf("Pod Name [%v] in the volume is not same as Pod Name [%v] in the storage class", PodName, podNameinSC), "is pod name in volume same as pod name in storage class?")
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:5116]: use dash verifications instead of log failures

File: tests/basic/pure_test.go [hunk 8]
```diff
@@ -5002,3 +5004,140 @@ var _ = Describe("{RebootingNodesWhileFADAvolumeCreationInProgressUsingZones}",
...
+			log.InfoD(stepLog)
+			for _, ctx := range contexts {
+				volumes, err := Inst().S.GetVolumes(ctx)
+				log.FailOnError(err, "Failed to get volumes")
+				for _, volume := range volumes {
+					inspectedVolume, err := Inst().V.InspectVolume(volume.ID)
+					log.FailOnError(err, "Failed to inspect volume")
+					PodName := inspectedVolume.Locator.VolumeLabels["pure_fa_pod_name"]
+					if PodName == "" {
+						log.FailOnError(fmt.Errorf("Pod Name is not present in the volume labels"), "is pod name present in volume labels?")
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:5114]: use dash verifications instead of log failures

File: tests/basic/pure_test.go [hunk 9]
```diff
@@ -5002,3 +5004,135 @@ var _ = Describe("{RebootingNodesWhileFADAvolumeCreationInProgressUsingZones}",
...
+				volumes, err := Inst().S.GetVolumes(ctx)
+				log.FailOnError(err, "Failed to get volumes")
+				for _, volume := range volumes {
+					inspectedVolume, err := Inst().V.InspectVolume(volume.ID)
+					log.FailOnError(err, "Failed to inspect volume")
+					PodName := inspectedVolume.Locator.VolumeLabels["pure_fa_pod_name"]
+					if PodName == "" {
+						dash.VerifyFatal(PodName == "", PodName != "", fmt.Sprintf("is pod name present in volume labels?"))
+					} else if PodName != podNameinSC {
+						dash.VerifyFatal(PodName != podNameinSC, PodName == podNameinSC, fmt.Sprintf("is pod name in volume same as pod name in storage class?"))
```
Comment (lsrinivas-pure) [inline tests/basic/pure_test.go:5111]: just this one check is enough 
dash.VerifyFatal(PodName , podNameinSC, "verify pod name in volume same as pod name in storage class")

**Pull request merged**